core/src/main/java/org/sonarsource/sonarlint/core/smartnotifications/SmartNotifications.java

```diff
@@ -0,0 +1,134 @@
+  private LastEventPolling lastEventPollingService;
+  private ScheduledExecutorService smartNotificationsPolling;
+
+  public SmartNotifications(ConfigurationRepository configurationRepository, ConnectionConfigurationRepository connectionRepository,
+    ServerApiProvider serverApiProvider, SonarLintClient client) {
+    this.configurationRepository = configurationRepository;
+    this.connectionRepository = connectionRepository;
+    this.serverApiProvider = serverApiProvider;
+    this.client = client;
+  }
+
+  public void initialize(Path storageRoot) {
+    lastEventPollingService = new LastEventPolling(storageRoot);
+    smartNotificationsPolling = Executors.newSingleThreadScheduledExecutor(r -> new Thread(r, "Smart Notifications Polling"));
+    smartNotificationsPolling.scheduleAtFixedRate(this::poll, 1, 60, TimeUnit.SECONDS);
+  }
+
+  private void poll() {
+    var keysAndScopeIdsPerConnectionId = getProjectKeysAndConfigScopeIdsPerConnectionIds();
+
+    for (var keysAndScopeIds : keysAndScopeIdsPerConnectionId.entrySet()) {
+      var connectionId = keysAndScopeIds.getKey();
+      var connection = connectionRepository.getEndpointParams(connectionId);
+      var httpClient = serverApiProvider.getServerApi(connectionId);
+
+      if (connection.isPresent() && httpClient.isPresent()) {
+        var serverApi = httpClient.get();
+        var developersApi = serverApi.developers();
+
+        if (developersApi.isSupported()) {
+          var projectKeysByLastEventPolling =
+            keysAndScopeIds.getValue().keySet().stream()
+              .distinct()
+              .collect(Collectors.toMap(Function.identity(),
+                p -> getLastNotificationTime(lastEventPollingService.getLastEventPolling(p, connectionId))));
+
+          var notifications = developersApi.getEvents(projectKeysByLastEventPolling)
+            .stream().map(e -> new ServerNotification(
+              e.getCategory(),
+              e.getMessage(),
+              e.getLink(),
+              e.getProjectKey(),
+              e.getTime()))
+            .collect(Collectors.toList());
+
+          for (var n : notifications) {
+            var smartNotification = new ShowSmartNotificationParams(n.message(), n.link(), keysAndScopeIds.getValue().get(n.projectKey()));
+            client.showSmartNotification(smartNotification);
+          }
+
+          projectKeysByLastEventPolling.keySet()
+            .forEach(projectKey -> lastEventPollingService.setLastEventPolling(ZonedDateTime.now(), projectKey, connectionId));
+        }
+      }
+    }
+  }
+
+  public void shutdown() {
+    smartNotificationsPolling.shutdownNow();
+  }
+
+  private static ZonedDateTime getLastNotificationTime(ZonedDateTime lastTime) {
+    var oneDayAgo = ZonedDateTime.now().minusDays(1);
+    return lastTime.isAfter(oneDayAgo) ? lastTime : oneDayAgo;
+  }
+
+  private Map<String, Map<String, Set<String>>> getProjectKeysAndConfigScopeIdsPerConnectionIds() {
```

Comment:
I suggest moving this method to ConfigurationRepository. Also it's a bit complex, what about something like:

```
public Map<String, Map<String, Set<String>>> getScopeIdsPerProjectKeyPerConnectionId() {
  Map<String, Map<String, Set<String>>> scopeIdsPerProjectKeyPerConnectionId = new HashMap<>();
  bindingPerConfigScopeId
    .forEach((scopeId, bindingConfiguration) -> scopeIdsPerProjectKeyPerConnectionId.computeIfAbsent(bindingConfiguration.getConnectionId(), k -> new HashMap<>())
      .computeIfAbsent(bindingConfiguration.getSonarProjectKey(), k -> new HashSet<>())
      .add(scopeId));
  return scopeIdsPerProjectKeyPerConnectionId;
}
```
Also the Map<String, Map<String, Set<String>>> looks primitive, there is probably a concept to introduce, like `ConnectionBindings`? To have a Map<String, ConnectionBindings>

Reply:
That's better thanks, it was just missing the check `isBound()` in your example

`ConnectionBindings` -> it would be more like `ProjectScopes` in our case, I'll try to come up with something convenient